java/src/org/openqa/selenium/grid/config/MapConfig.java

```diff
@@ -65,7 +68,24 @@ public Optional<List<String>> getAll(String section, String option) {
     }
 
     Object value = rawSection.get(option);
-    return value == null ? Optional.empty() : Optional.of(ImmutableList.of(String.valueOf(value)));
+    if (value == null) {
+      return Optional.empty();
+    }
+
+    if (value instanceof Collection<?>) {
+      Collection<?> collection = (Collection<?>) value;
+      // Case when an array of map is used as config
+      if (collection.stream().anyMatch(item -> item instanceof Map)) {
+        List<String> toReturn = collection.stream()
+          .map(item -> (Map<?, ?>) item)
+          .flatMap(map -> map.entrySet().stream())
+          .map(entry -> String.format("%s=%s", entry.getKey(), entry.getValue()))
```

Comment:
If the assumed structure is section/option/value, don't the values need to be encoded as JSON to avoid loss of fidelity?

Reply:
Alternatively, you could use **Map<String, Object>** as the internal representation across the board. I'm not sure which format provides the best transformation API, but it might be JSON.

Reply:
@sbabcoc - I am not quite sure what you are hinting at. Is there a sample json that you can share which you would like to be vetted out as part of this PR? I can check that out and if needed add it up as an additional test. I think that would be an easier way of answering your questions/concerns.

Reply:
@sbabcoc since you seem to be more closer to the problem, do you want to take over with this PR and wrap it to closure ?

Reply:
I'd be happy to! I just need to get Selenium to build on my Windows 10 machine. Can you point me to an up-to-date setup process?

Reply:
@sbabcoc The readme file has the instructions

Reply:
@krmahadevan I'll go through the instructions again, but my current setup fails with this:

```bash
zstd/lib/dictBuilder\zdict.c(53): fatal error C1083: Cannot open include file: '../compress/zstd_compress_internal.h': No such file or directory
```

Any hints of fixing this issue?

Reply:
It seems like this command, which is from one of the build scripts, might be missing in "include" declaration:

```bash
error occurred: Command "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/HostX64/x64/cl.exe" "-nologo" "-MD" "-Z7" "-Brepro" "/nologo" "/DCOMPILER_MSVC" "/DNOMINMAX" "/D_WIN32_WINNT=0x0601" "/D_CRT_SECURE_NO_DEPRECATE" "/D_CRT_SECURE_NO_WARNINGS" "/bigobj" "/Zm500" "/EHsc" "/wd4351" "/wd4291" "/wd4250" "/wd4996" "/showIncludes" "/MD" "/Od" "/Z7" "/wd4117" "-D__DATE__=\"redacted\"" "-D__TIMESTAMP__=\"redacted\"" "-D__TIME__=\"redacted\"" "-I" "zstd/lib/" "-I" "zstd/lib/common" "-I" "zstd/lib/legacy" "-fvisibility=hidden" "-DZSTD_DISABLE_ASM=" "-DZSTD_LIB_DEPRECATED=0" "-DXXH_PRIVATE_API=" "-DZSTDLIB_VISIBILITY=" "-DZDICTLIB_VISIBILITY=" "-DZSTDERRORLIB_VISIBILITY=" "-DZSTD_LEGACY_SUPPORT=1" "-FoC:\\users\\scoba\\_bazel_scoba\\io65vnbs\\execroot\\selenium\\bazel-out/x64_windows-fastbuild/bin/external/crates__zstd-sys-2.0.7-zstd.1.5.4/zstd-sys_build_script.out_dir\\zstd/lib/dictBuilder\\fastcover.o" "-c" "zstd/lib/dictBuilder\\fastcover.c" with args "cl.exe" did not execute successfully (status code exit code: 2).
```
The `zstd/lib/compress` folder appears to be absent. I haven't been able to track down where this command originates.

Reply:
@sbabcoc - I am not using Windows. So I am not sure what that error may be referring to. Maybe @titusfortner or @diemol can help here ?

Reply:
@sbabcoc From the slack channel, I found this gist that @titusfortner created https://gist.github.com/titusfortner/aec103e9b02709f771497fdb8b21154c Please see if that would help maybe ?

Reply:
@titusfortner I followed the build environment setup instructions @krmahadevan pointed out to me, but my Selenium build still fails. 

```
ERROR: C:/users/scoba/_bazel_scoba/io65vnbs/external/rules_rust/util/process_wrapper/BUILD.bazel:6:36: Compiling Rust (without process_wrapper) bin process_wrapper (6 files) [for tool] failed: (Exit 1): rustc.exe failed: error executing command (from target @rules_rust//util/process_wrapper:process_wrapper)
  cd /d C:/users/scoba/_bazel_scoba/io65vnbs/execroot/selenium
  SET CARGO_CFG_TARGET_ARCH=x86_64
    SET CARGO_CFG_TARGET_OS=windows
    SET CARGO_CRATE_NAME=process_wrapper
    SET CARGO_MANIFEST_DIR=${pwd}/external/rules_rust/util/process_wrapper
    SET CARGO_PKG_AUTHORS=
    SET CARGO_PKG_DESCRIPTION=
    SET CARGO_PKG_HOMEPAGE=
    SET CARGO_PKG_NAME=process_wrapper
    SET CARGO_PKG_VERSION=0.0.0
    SET CARGO_PKG_VERSION_MAJOR=0
    SET CARGO_PKG_VERSION_MINOR=0
    SET CARGO_PKG_VERSION_PATCH=0
    SET CARGO_PKG_VERSION_PRE=
    SET LIB=msvc_not_found
    SET PATH=msvc_not_found
    SET SYSROOT=bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools/rust_toolchain
    SET TEMP=msvc_not_found
    SET TMP=msvc_not_found
  bazel-out\x64_windows-opt-exec-2B5CBBC6\bin\external\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\rust_toolchain\bin\rustc.exe external/rules_rust/util/process_wrapper/main.rs --crate-name=process_wrapper --crate-type=bin --error-format=human --out-dir=bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper --codegen=opt-level=3 --codegen=debuginfo=0 --remap-path-prefix=${pwd}= --emit=dep-info,link --color=always --target=x86_64-pc-windows-msvc -L bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools/rust_toolchain/lib/rustlib/x86_64-pc-windows-msvc/lib --edition=2018 --codegen=linker=external/local_config_cc/vc_installation_error_x64.bat --codegen link-args=/nologo /SUBSYSTEM:CONSOLE advapi32.lib ws2_32.lib userenv.lib Bcrypt.lib /MACHINE:X64 /DEFAULTLIB:msvcrt.lib /OPT:ICF /OPT:REF --extern=tinyjson=bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust_tinyjson/libtinyjson-4031717389.rlib -Ldependency=bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust_tinyjson
# Configuration: 800883ccb1766b992fce1ea33e718d5c585d5508e44f48f7cd6e7bb86401d8b2
# Execution platform: @local_config_platform//:host
error: linking with `external/local_config_cc/vc_installation_error_x64.bat` failed: exit code: 1
  |
  = note: "cmd" "/c" "external/local_config_cc/vc_installation_error_x64.bat" "/NOLOGO" "C:\\Users\\scoba\\AppData\\Local\\Temp\\rustcaDwNye\\symbols.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.0.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.1.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.10.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.11.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.12.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.13.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.14.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.15.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.2.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.3.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.4.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.5.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.6.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.7.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.8.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.process_wrapper.848bfd2b-cgu.9.rcgu.o" "bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.3nzehrxjo1m8r9ix.rcgu.o" "/LIBPATH:bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools/rust_toolchain/lib/rustlib/x86_64-pc-windows-msvc/lib" "/LIBPATH:bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust_tinyjson" "/LIBPATH:C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\execroot\\selenium\\bazel-out\\x64_windows-opt-exec-2B5CBBC6\\bin\\external\\rules_rust_tinyjson\\libtinyjson-4031717389.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd-def8e53046de335b.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libpanic_unwind-62d160d4fb1f3e35.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_demangle-13ad55cee4080447.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd_detect-1735606bdcca0248.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libhashbrown-ed3da6b061bf16b2.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libminiz_oxide-798a382c9e175c85.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libadler-f40bb6e991434635.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-e0d1795fe60da182.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libunwind-05143401f2cd6cbf.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-d0fa84d9f351cf16.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liblibc-c00d1f51981f09f6.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc-6231dd8f826d6fe5.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_core-b7b22741583360f1.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-a8962116ce00a4da.rlib" "C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-58e2de43e9b1a814.rlib" "kernel32.lib" "advapi32.lib" "userenv.lib" "kernel32.lib" "ws2_32.lib" "bcrypt.lib" "msvcrt.lib" "legacy_stdio_definitions.lib" "/NXCOMPAT" "/LIBPATH:C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "/OUT:bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper\\process_wrapper.exe" "/OPT:REF,ICF" "/DEBUG" "/NATVIS:C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Users\\scoba\\_bazel_scoba\\io65vnbs\\external\\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\\lib\\rustlib\\etc\\libstd.natvis" "/nologo" "/SUBSYSTEM:CONSOLE" "advapi32.lib" "ws2_32.lib" "userenv.lib" "Bcrypt.lib" "/MACHINE:X64" "/DEFAULTLIB:msvcrt.lib" "/OPT:ICF" "/OPT:REF"
  = note: 'external' is not recognized as an internal or external command,
          operable program or batch file.


error: aborting due to previous error

Target //dotnet/test/common:chrome failed to build
ERROR: C:/github/selenium/java/src/org/openqa/selenium/netty/server/BUILD.bazel:4:13 Building java/src/org/openqa/selenium/netty/server/libserver.jar (11 source files) failed: (Exit 1): rustc.exe failed: error executing command (from target @rules_rust//util/process_wrapper:process_wrapper)
  cd /d C:/users/scoba/_bazel_scoba/io65vnbs/execroot/selenium
  SET CARGO_CFG_TARGET_ARCH=x86_64
    SET CARGO_CFG_TARGET_OS=windows
    SET CARGO_CRATE_NAME=process_wrapper
    SET CARGO_MANIFEST_DIR=${pwd}/external/rules_rust/util/process_wrapper
    SET CARGO_PKG_AUTHORS=
    SET CARGO_PKG_DESCRIPTION=
    SET CARGO_PKG_HOMEPAGE=
    SET CARGO_PKG_NAME=process_wrapper
    SET CARGO_PKG_VERSION=0.0.0
    SET CARGO_PKG_VERSION_MAJOR=0
    SET CARGO_PKG_VERSION_MINOR=0
    SET CARGO_PKG_VERSION_PATCH=0
    SET CARGO_PKG_VERSION_PRE=
    SET LIB=msvc_not_found
    SET PATH=msvc_not_found
    SET SYSROOT=bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools/rust_toolchain
    SET TEMP=msvc_not_found
    SET TMP=msvc_not_found
  bazel-out\x64_windows-opt-exec-2B5CBBC6\bin\external\rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools\rust_toolchain\bin\rustc.exe external/rules_rust/util/process_wrapper/main.rs --crate-name=process_wrapper --crate-type=bin --error-format=human --out-dir=bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust/util/process_wrapper --codegen=opt-level=3 --codegen=debuginfo=0 --remap-path-prefix=${pwd}= --emit=dep-info,link --color=always --target=x86_64-pc-windows-msvc -L bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rust_windows_x86_64__x86_64-pc-windows-msvc__stable_tools/rust_toolchain/lib/rustlib/x86_64-pc-windows-msvc/lib --edition=2018 --codegen=linker=external/local_config_cc/vc_installation_error_x64.bat --codegen link-args=/nologo /SUBSYSTEM:CONSOLE advapi32.lib ws2_32.lib userenv.lib Bcrypt.lib /MACHINE:X64 /DEFAULTLIB:msvcrt.lib /OPT:ICF /OPT:REF --extern=tinyjson=bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust_tinyjson/libtinyjson-4031717389.rlib -Ldependency=bazel-out/x64_windows-opt-exec-2B5CBBC6/bin/external/rules_rust_tinyjson
```
The setting whose values show `msvc_not_found` are certainly indicative of configuration problems. Suggestions?

Reply:
@krmahadevan Thanks for the pointer! That was very helpful, although I mis-interpreted one of the instructions. I now run into the same issue I had when I started:

```
zstd/lib/dictBuilder\zdict.c(53): fatal error C1083: Cannot open include file: '../compress/zstd_compress_internal.h': No such file or directory
```